File: frontend/src/services/webrtcService.js
import api from './api';

class WebRTCService {
    constructor() {
        this.peerConnections = new Map();
        this.localStream = null;
        this.remoteStreams = new Map();
        this.roomId = null;
        this.userId = null;
        this.onMessageCallback = null;
        this.onParticipantUpdateCallback = null;
        this.onRemoteStreamCallback = null;
        this.onParticipantLeftCallback = null;
        this.webSocket = null;
        this.isConnected = false;
        
        // WebRTC configuration
        this.rtcConfig = {
            iceServers: [
                { urls: 'stun:stun.l.google.com:19302' },
                { urls: 'stun:stun1.l.google.com:19302' }
            ]
        };
    }

    // Event handlers
    onParticipantUpdate(callback) {
        this.onParticipantUpdateCallback = callback;
    }

    onRemoteStream(callback) {
        this.onRemoteStreamCallback = callback;
    }

    onParticipantLeft(callback) {
        this.onParticipantLeftCallback = callback;
    }

    onMessage(callback) {
        this.onMessageCallback = callback;
    }

    // WebSocket connection
    async connectWebSocket() {
        try {
            const token = localStorage.getItem('authToken');
            const wsUrl = `ws://localhost:8001/ws/interview/${this.roomId}/?token=${token}`;
            
            this.webSocket = new WebSocket(wsUrl);
            
            this.webSocket.onopen = () => {
                console.log('🔌 WebSocket connected');
                this.isConnected = true;
            };
            
            this.webSocket.onmessage = (event) => {
                this.handleWebSocketMessage(JSON.parse(event.data));
            };
            
            this.webSocket.onclose = () => {
                console.log('🔌 WebSocket disconnected');
                this.isConnected = false;
            };
            
            this.webSocket.onerror = (error) => {
                console.error('WebSocket error:', error);
            };
        } catch (error) {
            console.error('Error connecting WebSocket:', error);
        }
    }

    handleWebSocketMessage(data) {
        console.log('📨 WebSocket message received:', data);
        
        switch (data.type) {
            case 'connection_established':
                console.log('✅ WebSocket connection established');
                break;
            case 'user_joined':
                console.log('👤 User joined message received:', data);
                this.handleParticipantJoined(data);
                break;
            case 'user_left':
                console.log('👤 User left message received:', data);
                this.handleParticipantLeft(data);
                break;
            case 'webrtc_signal':
                this.handleWebRTCSignal(data);
                break;
            case 'chat_message':
                if (this.onMessageCallback) {
                    this.onMessageCallback(data);
                }
                break;
            case 'participants_updated':
                console.log('👥 Participants updated message received:', data);
                console.log('👥 Participants data:', data.participants);
                console.log('👥 Callback available:', !!this.onParticipantUpdateCallback);
                if (this.onParticipantUpdateCallback) {
                    this.onParticipantUpdateCallback(data.participants);
                } else {
                    console.warn('⚠️ No participant update callback set');
                }
                break;
            default:
                console.log('Unknown message type:', data.type);
        }
    }

    sendWebSocketMessage(type, payload) {
        if (this.webSocket && this.webSocket.readyState === WebSocket.OPEN) {
            const message = { type, ...payload };
            console.log('📤 Sending WebSocket message:', message);
            this.webSocket.send(JSON.stringify(message));
        } else {
            console.warn('⚠️ WebSocket not ready, message not sent:', type);
        }
    }

    // Room Management
    async createRoom(interviewId, roomConfig = {}) {
        try {
            const response = await api.post('/interview/webrtc/create-room/', {
                interview_id: interviewId,
                room_config: roomConfig
            });
            return response.data;
        } catch (error) {
            console.error('Error creating room:', error);
            throw error;
        }
    }

    async joinRoom(roomId, participantType = 'candidate') {
        try {
            console.log('🚪 Joining room:', roomId);
            
            const response = await api.post('/interview/webrtc/join-room/', {
                room_id: roomId,
                participant_type: participantType
            });
            
            this.roomId = roomId;
            this.userId = response.data.participant.user;
            
            // Connect WebSocket
            await this.connectWebSocket();
            
                            // Wait for WebSocket connection
                await new Promise((resolve) => {
                    const checkConnection = () => {
                        if (this.webSocket && this.webSocket.readyState === WebSocket.OPEN) {
                            resolve();
                        } else {
                            setTimeout(checkConnection, 100);
                        }
                    };
                    checkConnection();
                });
                
                // Send join room message via WebSocket
                this.sendWebSocketMessage('join_room', {
                    participant_type: participantType
                });
            
            // Initialize WebRTC
            await this.initializeWebRTC();
            
            console.log('✅ Successfully joined room');
            return response.data;
        } catch (error) {
            console.error('Error joining room:', error);
            throw error;
        }
    }

    async leaveRoom() {
        try {
            if (this.roomId) {
                await api.post('/interview/webrtc/leave-room/', {
                    room_id: this.roomId
                });
                
                // Clean up WebRTC connections
                this.cleanup();
                
                // Close WebSocket
                if (this.webSocket) {
                    this.webSocket.close();
                }
            }
        } catch (error) {
            console.error('Error leaving room:', error);
        }
    }

    // Set local stream
    setLocalStream(stream) {
        this.localStream = stream;
        console.log('📹 Local stream set in webrtcService');
    }

    // WebRTC Initialization
    async initializeWebRTC() {
        try {
            // Note: Local stream should be set by the calling component
            // This method is now just for initialization setup
            console.log('📹 WebRTC initialized');
        } catch (error) {
            console.error('Error initializing WebRTC:', error);
            throw error;
        }
    }

    // WebRTC Signaling
    async handleWebRTCSignal(signalData) {
        const { from_user_id, signal_type, signal_data } = signalData;
        
        try {
            let peerConnection = this.peerConnections.get(from_user_id);
            
            if (!peerConnection) {
                peerConnection = this.createPeerConnection(from_user_id);
            }
            
            switch (signal_type) {
                case 'offer':
                    // Only handle offer if we don't already have a remote description
                    if (peerConnection.remoteDescription === null) {
                        await peerConnection.setRemoteDescription(new RTCSessionDescription(signal_data));
                        const answer = await peerConnection.createAnswer();
                        await peerConnection.setLocalDescription(answer);
                        
                        this.sendWebSocketMessage('webrtc_signal', {
                            target_user_id: from_user_id,
                            signal_type: 'answer',
                            signal_data: answer
                        });
                    }
                    break;
                    
                case 'answer':
                    // Only handle answer if we have a local description and no remote description
                    if (peerConnection.localDescription && peerConnection.remoteDescription === null) {
                        await peerConnection.setRemoteDescription(new RTCSessionDescription(signal_data));
                    }
                    break;
                    
                case 'ice_candidate':
                    // Only add ICE candidate if we have a remote description
                    if (peerConnection.remoteDescription) {
                        await peerConnection.addIceCandidate(new RTCIceCandidate(signal_data));
                    }
                    break;
            }
        } catch (error) {
            console.error('Error handling WebRTC signal:', error);
        }
    }

    createPeerConnection(remoteUserId) {
        console.log('🔗 Creating peer connection with:', remoteUserId, 'Current user:', this.userId);
        
        // Don't create peer connection with ourselves
        if (remoteUserId === this.userId) {
            console.log('🔗 Skipping peer connection with self');
            return null;
        }
        
        const peerConnection = new RTCPeerConnection(this.rtcConfig);
        
        // Add local stream
        if (this.localStream) {
            this.localStream.getTracks().forEach(track => {
                peerConnection.addTrack(track, this.localStream);
            });
        }
        
        // Handle remote stream
        peerConnection.ontrack = (event) => {
            console.log('📹 Remote stream received from:', remoteUserId);
            console.log('📹 Stream details:', event.streams[0]);
            this.remoteStreams.set(remoteUserId, event.streams[0]);
            
            if (this.onRemoteStreamCallback) {
                console.log('📹 Calling remote stream callback for:', remoteUserId);
                this.onRemoteStreamCallback(remoteUserId, event.streams[0]);
            } else {
                console.warn('⚠️ No remote stream callback set');
            }
        };
        
        // Handle ICE candidates
        peerConnection.onicecandidate = (event) => {
            if (event.candidate) {
                this.sendWebSocketMessage('webrtc_signal', {
                    target_user_id: remoteUserId,
                    signal_type: 'ice_candidate',
                    signal_data: event.candidate
                });
            }
        };
        
        // Handle connection state changes
        peerConnection.onconnectionstatechange = () => {
            console.log('Connection state:', peerConnection.connectionState);
        };
        
        this.peerConnections.set(remoteUserId, peerConnection);
        return peerConnection;
    }

    async handleParticipantJoined(participant) {
        console.log('👤 Participant joined:', participant.user_id);
        console.log('👤 Current user ID:', this.userId);
        
        // Don't create peer connection with ourselves
        if (participant.user_id === this.userId) {
            console.log('👤 Skipping peer connection with self');
            return;
        }
        
        // Create offer for new participant
        const peerConnection = this.createPeerConnection(participant.user_id);
        if (!peerConnection) {
            console.log('👤 No peer connection created (self-connection prevented)');
            return;
        }
        const offer = await peerConnection.createOffer();
        await peerConnection.setLocalDescription(offer);
        
        this.sendWebSocketMessage('webrtc_signal', {
            target_user_id: participant.user_id,
            signal_type: 'offer',
            signal_data: offer
        });
    }

    handleParticipantLeft(participant) {
        console.log('👤 Participant left:', participant.user_id);
        
        // Clean up peer connection
        const peerConnection = this.peerConnections.get(participant.user_id);
        if (peerConnection) {
            peerConnection.close();
            this.peerConnections.delete(participant.user_id);
        }
        
        // Remove remote stream
        this.remoteStreams.delete(participant.user_id);
        
        if (this.onParticipantLeftCallback) {
            this.onParticipantLeftCallback(participant.user_id);
        }
    }

    // Utility methods
    getLocalStream() {
        return this.localStream;
    }

    getVideoSender() {
        // Get the first video sender from any peer connection
        for (const [userId, pc] of this.peerConnections) {
            const sender = pc.getSenders().find(s => s.track?.kind === 'video');
            if (sender) return sender;
        }
        return null;
    }

    // Chat Messages
    async sendMessage(message) {
        try {
            this.sendWebSocketMessage('chat_message', {
                message: message,
                room_id: this.roomId
            });
            
            return {
                id: Date.now(),
                message: message,
                user_id: this.userId,
                timestamp: new Date().toISOString()
            };
        } catch (error) {
            console.error('Error sending message:', error);
            throw error;
        }
    }

    async getMessages(limit = 50) {
        try {
            const response = await api.get(`/interview/webrtc/messages/${this.roomId}/?limit=${limit}`);
            return response.data.messages || [];
        } catch (error) {
            console.error('Error fetching messages:', error);
            return [];
        }
    }

    // Cleanup
    cleanup() {
        // Close all peer connections
        this.peerConnections.forEach(pc => pc.close());
        this.peerConnections.clear();
        
        // Stop local stream
        if (this.localStream) {
            this.localStream.getTracks().forEach(track => track.stop());
            this.localStream = null;
        }
        
        // Clear remote streams
        this.remoteStreams.clear();
        
        // Close WebSocket
        if (this.webSocket) {
            this.webSocket.close();
            this.webSocket = null;
        }
        
        this.isConnected = false;
        this.roomId = null;
        this.userId = null;
    }
}

export default new WebRTCService();

File: backend/interview_management/webrtc_service.py
import json
import logging
from typing import Dict, List, Optional
from django.conf import settings
from .models import InterviewRoom, RoomParticipant, ChatMessage, InterviewSession
from django.contrib.auth.models import User
from channels.layers import get_channel_layer
from asgiref.sync import async_to_sync

logger = logging.getLogger(__name__)

class WebRTCService:
    """Service for managing WebRTC connections and signaling"""
    
    def __init__(self):
        try:
            self.channel_layer = get_channel_layer()
        except:
            self.channel_layer = None
        self.ice_servers = [
            {"urls": "stun:stun.l.google.com:19302"},
            {"urls": "stun:stun1.l.google.com:19302"},
        ]
    
    def create_room(self, interview_id: str, room_config: Dict = None) -> InterviewRoom:
        """Create a new interview room"""
        try:
            interview = InterviewSession.objects.get(id=interview_id)
            
            room_id = f"room_{interview_id}_{interview.created_at.strftime('%Y%m%d_%H%M%S')}"
            
            room = InterviewRoom.objects.create(
                room_id=room_id,
                interview=interview,
                ice_servers=self.ice_servers,
                recording_enabled=room_config.get('recording_enabled', True) if room_config else True,
                screen_sharing_enabled=room_config.get('screen_sharing_enabled', True) if room_config else True,
                chat_enabled=room_config.get('chat_enabled', True) if room_config else True,
            )
            
            logger.info(f"Created interview room: {room_id}")
            return room
            
        except Exception as e:
            logger.error(f"Error creating room: {e}")
            raise
    
    def join_room(self, room_id: str, user_id: int, participant_type: str, peer_id: str = None) -> RoomParticipant:
        """Join an interview room"""
        try:
            room = InterviewRoom.objects.get(room_id=room_id, is_active=True)
            user = User.objects.get(id=user_id)
            
            # Check if user is already a participant
            participant, created = RoomParticipant.objects.get_or_create(
                room=room,
                user=user,
                defaults={
                    'participant_type': participant_type,
                    'peer_id': peer_id,
                    'connection_state': 'connecting'
                }
            )
            
            if not created:
                # Update existing participant
                participant.is_active = True
                participant.connection_state = 'connecting'
                if peer_id:
                    participant.peer_id = peer_id
                participant.save()
            
            # Send join notification to other participants
            self._notify_participants(room_id, 'user_joined', {
                'user_id': user_id,
                'user_email': user.email,
                'participant_type': participant_type,
                'peer_id': peer_id
            })
            
            logger.info(f"User {user.email} joined room {room_id}")
            return participant
            
        except Exception as e:
            logger.error(f"Error joining room: {e}")
            raise
    
    def leave_room(self, room_id: str, user_id: int) -> bool:
        """Leave an interview room"""
        try:
            participant = RoomParticipant.objects.get(
                room__room_id=room_id,
                user_id=user_id,
                is_active=True
            )
            
            participant.leave_room()
            
            # Send leave notification to other participants
            self._notify_participants(room_id, 'user_left', {
                'user_id': user_id,
                'user_email': participant.user.email,
                'peer_id': participant.peer_id
            })
            
            logger.info(f"User {participant.user.email} left room {room_id}")
            return True
            
        except RoomParticipant.DoesNotExist:
            logger.warning(f"User {user_id} not found in room {room_id}")
            return False
        except Exception as e:
            logger.error(f"Error leaving room: {e}")
            raise
    
    def send_offer(self, room_id: str, from_user_id: int, to_user_id: int, offer: Dict) -> bool:
        """Send WebRTC offer"""
        try:
            self._notify_user(room_id, to_user_id, 'webrtc_offer', {
                'from_user_id': from_user_id,
                'offer': offer
            })
            return True
        except Exception as e:
            logger.error(f"Error sending offer: {e}")
            return False
    
    def send_answer(self, room_id: str, from_user_id: int, to_user_id: int, answer: Dict) -> bool:
        """Send WebRTC answer"""
        try:
            self._notify_user(room_id, to_user_id, 'webrtc_answer', {
                'from_user_id': from_user_id,
                'answer': answer
            })
            return True
        except Exception as e:
            logger.error(f"Error sending answer: {e}")
            return False
    
    def send_ice_candidate(self, room_id: str, from_user_id: int, to_user_id: int, candidate: Dict) -> bool:
        """Send ICE candidate"""
        try:
            self._notify_user(room_id, to_user_id, 'ice_candidate', {
                'from_user_id': from_user_id,
                'candidate': candidate
            })
            return True
        except Exception as e:
            logger.error(f"Error sending ICE candidate: {e}")
            return False
    
    def send_chat_message(self, room_id: str, user_id: int, message: str) -> ChatMessage:
        """Send a chat message"""
        try:
            room = InterviewRoom.objects.get(room_id=room_id)
            user = User.objects.get(id=user_id)
            
            chat_message = ChatMessage.objects.create(
                room=room,
                sender=user,
                message=message
            )
            
            # Notify all participants about the message
            self._notify_participants(room_id, 'chat_message', {
                'message_id': str(chat_message.id),
                'user_id': user_id,
                'user_email': user.email,
                'message': message,
                'timestamp': chat_message.timestamp.isoformat()
            })
            
            return chat_message
            
        except Exception as e:
            logger.error(f"Error sending chat message: {e}")
            raise
    
    def get_room_participants(self, room_id: str) -> List[Dict]:
        """Get all active participants in a room"""
        try:
            participants = RoomParticipant.objects.filter(
                room__room_id=room_id,
                is_active=True
            ).select_related('user')
            
            return [{
                'user_id': p.user.id,
                'user_email': p.user.email,
                'participant_type': p.participant_type,
                'peer_id': p.peer_id,
                'connection_state': p.connection_state,
                'joined_at': p.joined_at.isoformat()
            } for p in participants]
            
        except Exception as e:
            logger.error(f"Error getting room participants: {e}")
            return []
    
    def get_room_messages(self, room_id: str, limit: int = 50) -> List[Dict]:
        """Get recent chat messages for a room"""
        try:
            messages = ChatMessage.objects.filter(
                room__room_id=room_id
            ).select_related('sender').order_by('-timestamp')[:limit]
            
            return [{
                'message_id': str(m.id),
                'user_id': m.sender.id,
                'user_email': m.sender.email,
                'message': m.message,
                'timestamp': m.timestamp.isoformat(),
                'is_system_message': m.is_system_message
            } for m in reversed(messages)]  # Return in chronological order
            
        except Exception as e:
            logger.error(f"Error getting room messages: {e}")
            return []
    
    def _notify_participants(self, room_id: str, event_type: str, data: Dict):
        """Notify all participants in a room"""
        if not self.channel_layer:
            logger.warning("Channel layer not available, skipping notification")
            return
        try:
            async_to_sync(self.channel_layer.group_send)(
                f"room_{room_id}",
                {
                    "type": "interview.message",
                    "event_type": event_type,
                    "data": data
                }
            )
        except Exception as e:
            logger.error(f"Error notifying participants: {e}")
    
    def _notify_user(self, room_id: str, user_id: int, event_type: str, data: Dict):
        """Notify a specific user in a room"""
        if not self.channel_layer:
            logger.warning("Channel layer not available, skipping notification")
            return
        try:
            async_to_sync(self.channel_layer.group_send)(
                f"room_{room_id}_user_{user_id}",
                {
                    "type": "interview.message",
                    "event_type": event_type,
                    "data": data
                }
            )
        except Exception as e:
            logger.error(f"Error notifying user: {e}")

# Global WebRTC service instance
webrtc_service = WebRTCService()
